src/pybind/mgr/diskprediction/task.py

```diff
@@ -0,0 +1,163 @@
+        self._obj_sender = None
+        self._start_time = None
+        self._th = None
+
+        self.exit = False
+        self.event = Event()
+        self.task_interval = \
+            self._context.get_configuration(self._interval_key)
+        self.cluster_domain_id = \
+            self._context.get_configuration('diskprediction_cluster_domain_id')
+
+    def run(self):
+        self._start_time = time.time()
+        # self._run()
+        self._th = Thread(target=self._start, args=(self,))
+        self._th.start()
+
+    def terminate(self):
+        self.exit = True
+        self.event.set()
+        if self._th:
+            self._th.join(60)
+        self._log.info("PDS terminate %s complete" % self._task_name)
+
+    @staticmethod
+    def _start(obj_task):
+        obj_task._log.debug(
+            "start %s, interval: %s"
+            % (obj_task._task_name, obj_task.task_interval))
+        while not obj_task.exit:
+            obj_task.run_agents()
+            if obj_task.event:
+                obj_task.event.wait(int(obj_task.task_interval))
+                obj_task.event.clear()
+        obj_task._log.info(
+            "completed %s(%s)"
+            % (obj_task._task_name, time.time()-obj_task._start_time))
+
+    def run_agents(self):
+        # for testing
+        try:
+            self._log.debug("run_agents %s" % self._task_name)
+            self._obj_sender = Command(
+                host=self._context.get_configuration("diskprediction_server"),
+                user=self._context.get_configuration("diskprediction_user"),
+                password=self._context.get_configuration("diskprediction_password"),
+                port=self._context.get_configuration("diskprediction_port"),
+                dbname=self._context.get_configuration('diskprediction_database'),
+                context=self._context)
+            if not self._obj_sender:
+                self._log.error("invalid diskprediction sender")
+                self._context.status = DP_MGR_STAT_FAILED
+                return
+            if self._obj_sender.test_connection():
+                self._log.debug("succeed to test connection")
+                self._run()
+                self._context.status = DP_MGR_STAT_OK
+            else:
+                self._log.error("failed to test connection")
+                self._context.status = DP_MGR_STAT_FAILED
+        except Exception as e:
+            self._context.status = DP_MGR_STAT_FAILED
+            self._log.error(
+                "failed to start %s agents, %s" % (self._task_name, str(e)))
+
+    def _run(self):
+        pass
+
+
+class MetricsTask(DPTask):
+
+    _task_name = "Metrics Task"
+    _interval_key = "diskprediction_upload_metrics_interval"
+    _agents = [CephCluster_Agent, CephMon_Agent, CephOSD_Agent, CephPool_Agent,
+               SAI_CluserAgent, SAI_DiskAgent, SAI_HostAgent, DB_RelayAgent,
+               SAI_AgentAgent]
+
+    def _run(self):

```

Comment:
This run method appears to be copy-pasted between these classes.  Move it up into the definition of DPTask.